File: frontend/src/api/jobs.ts
const BASE_URL = (import.meta.env.VITE_API_URL as string | undefined) ?? 'http://localhost:8000'

export interface InterviewRoundType {
  id: number
  name: string
  description: string | null
}

export interface InterviewRoundInput {
  round_type_id: number
  round_order: number
  failing_criteria: number | null
  description?: string
}

export interface ATSCriterionInput {
  section: string
  weight: number
}

export interface JobInterviewRoundItem {
  round_order: number
  round_type_name: string
  failing_criteria: number | null
  description: string | null
}

export interface ATSCriterionSummary {
  section: string
  weight: number
}

export interface ATSCriteriaSummary {
  has_config: boolean
  criteria: ATSCriterionSummary[]
  qualify_threshold: number
  overqualify_threshold: number | null
  auto_reject_overqualified: boolean
}

export interface JobListItem {
  id: string
  description: string
  company: string
  job_role_id: number
  job_role_title: string
  experience_level_id: number
  experience_level_name: string
  location: string
  job_type: string
  salary_range: string | null
  posted_at: string
  expires_at: string | null
  required_skills: string[]
  status: string
  // Only populated on recruiter-facing listings (fetchRecruiterJobs) — null for the
  // public/candidate job list.
  ats_criteria: ATSCriteriaSummary | null
}

// ── Analytics types ───────────────────────────────────────────────────────────

export interface JobStatsRound {
  round_order: number
  round_type_name: string
  failing_criteria: number | null
  applicants_count: number
}

export interface JobStatsResponse {
  job_title: string
  description: string
  status: string
  required_skills: string[]
  rounds: JobStatsRound[]
  total_applicants: number
  passed_all_rounds: number
  hired_count: number
}

export interface RoundCandidateItem {
  candidate_id: string
  application_id: string
  interview_id: string
  name: string
  status: string
}

export interface CandidateSkillItem {
  name: string
  proficiency_level: string | null
}

export interface CandidateInfo {
  first_name: string
  last_name: string
  email: string
  bio: string | null
  current_location: string | null
  experience_level: string | null
  job_role: string | null
  skills: CandidateSkillItem[]
  phone: string | null
  linkedin_url: string | null
  experience_years_min: number | null
  experience_years_max: number | null
}

export interface InterviewProgressItem {
  round_order: number
  round_type_name: string
  status: string | null
  result: number | null
  completed_at: string | null
  interview_id: string | null
}

export interface EvaluationQuestionItem {
  question_text: string
  candidate_answer: string | null
  score: number | null
  notes: string | null
}

export interface CandidatePanelResponse {
  candidate: CandidateInfo
  progress: InterviewProgressItem[]
  evaluation: EvaluationQuestionItem[] | null
}

export interface JobFilters {
  job_role_id?: number
  experience_level_id?: number
  location?: string
  job_type?: string
  salary_range?: string
  candidate_id?: string
  offset?: number
  limit?: number
}

export interface JobPostRequest {
  recruiter_id: string
  job_role_id: number
  experience_level_id: number
  description: string
  location: string
  job_type: string
  salary_range?: string
  expires_at?: string
  skill_ids?: number[]
  interview_rounds?: InterviewRoundInput[]
  ats_criteria: ATSCriterionInput[]
  qualify_threshold?: number
  overqualify_threshold?: number
  auto_reject_overqualified?: boolean
}

export interface JobPostResponse {
  job_id: string
  message: string
}

export interface JobUpdateRequest {
  description?: string
  location?: string
  job_type?: string
  salary_range?: string | null
  expires_at?: string | null
  status?: 'active' | 'closed'
  skill_ids?: number[]
  ats_criteria?: ATSCriterionInput[]
  qualify_threshold?: number
  overqualify_threshold?: number | null
  auto_reject_overqualified?: boolean
}

export interface JobSkillOptionItem {
  id: number
  name: string
}

export interface RerunAtsResponse {
  queued: number
  skipped_pending: number
  excluded: number
  message: string
}

export interface RerunAtsStatusResponse {
  total_queued: number
  completed: number
  in_progress: boolean
}

export async function fetchInterviewRoundTypes(): Promise<InterviewRoundType[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/round-types`)
  if (!res.ok) throw new Error('Failed to load interview round types')
  return res.json() as Promise<InterviewRoundType[]>
}

export async function fetchJobRounds(jobId: string): Promise<JobInterviewRoundItem[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/rounds`)
  if (!res.ok) throw new Error('Failed to load interview rounds')
  return res.json() as Promise<JobInterviewRoundItem[]>
}

export async function fetchJobs(
  filters: JobFilters = {},
  signal?: AbortSignal,
): Promise<JobListItem[]> {
  const params = new URLSearchParams()
  if (filters.job_role_id != null) params.set('job_role_id', String(filters.job_role_id))
  if (filters.experience_level_id != null) params.set('experience_level_id', String(filters.experience_level_id))
  if (filters.location) params.set('location', filters.location)
  if (filters.job_type) params.set('job_type', filters.job_type)
  if (filters.salary_range) params.set('salary_range', filters.salary_range)
  if (filters.candidate_id) params.set('candidate_id', filters.candidate_id)
  if (filters.offset != null) params.set('offset', String(filters.offset))
  if (filters.limit != null) params.set('limit', String(filters.limit))
  const query = params.toString()
  const res = await fetch(`${BASE_URL}/api/jobs${query ? `?${query}` : ''}`, { signal })
  if (!res.ok) throw new Error('Failed to load jobs')
  return res.json() as Promise<JobListItem[]>
}

export async function fetchRecruiterJobs(recruiterId: string): Promise<JobListItem[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/mine?recruiter_id=${encodeURIComponent(recruiterId)}`)
  if (!res.ok) throw new Error('Failed to load your job postings')
  return res.json() as Promise<JobListItem[]>
}

export async function fetchInterviewQA(interviewId: string): Promise<EvaluationQuestionItem[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/interview-qa/${encodeURIComponent(interviewId)}`)
  if (!res.ok) throw new Error('Failed to load interview Q&A')
  return res.json() as Promise<EvaluationQuestionItem[]>
}

export async function fetchJobStats(jobId: string): Promise<JobStatsResponse> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/stats`)
  if (!res.ok) throw new Error('Failed to load job stats')
  return res.json() as Promise<JobStatsResponse>
}

export async function fetchRoundCandidates(jobId: string, roundOrder: number): Promise<RoundCandidateItem[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/rounds/${roundOrder}/candidates`)
  if (!res.ok) throw new Error('Failed to load round candidates')
  return res.json() as Promise<RoundCandidateItem[]>
}

export async function fetchCandidatePanel(
  jobId: string,
  applicationId: string,
  interviewId: string,
): Promise<CandidatePanelResponse> {
  const res = await fetch(
    `${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/candidate-panel/${encodeURIComponent(applicationId)}?interview_id=${encodeURIComponent(interviewId)}`,
  )
  if (!res.ok) throw new Error('Failed to load candidate details')
  return res.json() as Promise<CandidatePanelResponse>
}

export async function postJob(data: JobPostRequest): Promise<JobPostResponse> {
  const res = await fetch(`${BASE_URL}/api/jobs`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
  const body = (await res.json()) as { detail?: string } & Partial<JobPostResponse>
  if (!res.ok) throw new Error(body.detail ?? 'Failed to post job')
  return body as JobPostResponse
}

export async function fetchJobSkills(jobId: string): Promise<JobSkillOptionItem[]> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/skills`)
  if (!res.ok) throw new Error('Failed to load job skills')
  return res.json() as Promise<JobSkillOptionItem[]>
}

export async function updateJob(jobId: string, recruiterId: string, data: JobUpdateRequest): Promise<JobPostResponse> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}?recruiter_id=${encodeURIComponent(recruiterId)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(data),
  })
  const body = (await res.json()) as { detail?: string } & Partial<JobPostResponse>
  if (!res.ok) throw new Error(body.detail ?? 'Failed to update job')
  return body as JobPostResponse
}

export async function rerunAts(jobId: string, recruiterId: string): Promise<RerunAtsResponse> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/rerun-ats?recruiter_id=${encodeURIComponent(recruiterId)}`, {
    method: 'POST',
  })
  const body = (await res.json()) as { detail?: string } & Partial<RerunAtsResponse>
  if (!res.ok) throw new Error(body.detail ?? 'Failed to start ATS rerun')
  return body as RerunAtsResponse
}

export async function fetchAtsRerunStatus(jobId: string): Promise<RerunAtsStatusResponse> {
  const res = await fetch(`${BASE_URL}/api/jobs/${encodeURIComponent(jobId)}/rerun-ats/status`)
  if (!res.ok) throw new Error('Failed to load ATS rerun status')
  return res.json() as Promise<RerunAtsStatusResponse>
}

File: backend/app/schemas/jobs.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator

VALID_JOB_TYPES = ('Full-time', 'Part-time', 'Remote', 'Contract', 'Hybrid')


class InterviewRoundTypeItem(BaseModel):
    id: int
    name: str
    description: str | None


class InterviewRoundInput(BaseModel):
    round_type_id: int
    round_order: int
    failing_criteria: int | None = None  # 0–100 percent
    description: str | None = None
    time_limit_minutes: int | None = None  # written-test time limit; falls back to a global default when unset

    @field_validator('failing_criteria')
    @classmethod
    def validate_failing_criteria(cls, v: int | None) -> int | None:
        if v is not None and not (0 <= v <= 100):
            raise ValueError('failing_criteria must be between 0 and 100')
        return v

    @field_validator('time_limit_minutes')
    @classmethod
    def validate_time_limit_minutes(cls, v: int | None) -> int | None:
        if v is not None and v <= 0:
            raise ValueError('time_limit_minutes must be positive')
        return v


class ATSCriterionInput(BaseModel):
    section: Literal['experience', 'skills', 'projects', 'certifications', 'education', 'achievements']
    weight: float = Field(ge=0, le=100)


class JobPostRequest(BaseModel):
    recruiter_id: str
    job_role_id: int
    experience_level_id: int
    description: str
    location: str
    job_type: Literal['Full-time', 'Part-time', 'Remote', 'Contract', 'Hybrid']
    salary_range: str | None = None
    expires_at: str | None = None
    skill_ids: list[int] = []
    interview_rounds: list[InterviewRoundInput] = []
    ats_criteria: list[ATSCriterionInput] = Field(..., min_length=1)
    # Axis 1 (PASS/FAIL) — weighted_average vs this threshold. Defaults to 65 when not set by the
    # recruiter, rather than being required, so the form doesn't force every recruiter to think
    # about it.
    qualify_threshold: float = Field(65, ge=0, le=100)
    # Axis 2 (overqualification flag) — independent of qualify_threshold. None means
    # overqualification is never auto-rejected, only shown informationally.
    overqualify_threshold: float | None = Field(None, ge=0)
    auto_reject_overqualified: bool = False

    @field_validator('description', 'location', 'recruiter_id')
    @classmethod
    def validate_nonempty(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError('This field is required')
        return v

    @field_validator('ats_criteria')
    @classmethod
    def validate_ats_criteria(cls, v: list[ATSCriterionInput]) -> list[ATSCriterionInput]:
        if abs(sum(c.weight for c in v) - 100) > 1e-6:
            raise ValueError('ats_criteria weights must sum to 100')
        return v

    @field_validator('salary_range')
    @classmethod
    def validate_salary_range(cls, v: str | None) -> str | None:
        if v is not None:
            v = v.strip()
            if not v:
                return None
        return v


class JobUpdateRequest(BaseModel):
    """Recruiter edit of an already-posted job. Deliberately excludes `interview_rounds` —
    rounds are locked once a job is live (candidates may already have Interviews rows tied
    to them); everything else is editable."""
    description: str | None = None
    location: str | None = None
    job_type: Literal['Full-time', 'Part-time', 'Remote', 'Contract', 'Hybrid'] | None = None
    salary_range: str | None = None
    expires_at: str | None = None
    status: Literal['active', 'closed'] | None = None
    skill_ids: list[int] | None = None
    ats_criteria: list[ATSCriterionInput] | None = None
    qualify_threshold: float | None = Field(None, ge=0, le=100)
    overqualify_threshold: float | None = Field(None, ge=0)
    auto_reject_overqualified: bool | None = None

    @field_validator('ats_criteria')
    @classmethod
    def validate_ats_criteria(cls, v: list[ATSCriterionInput] | None) -> list[ATSCriterionInput] | None:
        if v is not None and abs(sum(c.weight for c in v) - 100) > 1e-6:
            raise ValueError('ats_criteria weights must sum to 100')
        return v


class JobPostResponse(BaseModel):
    job_id: str
    message: str


class JobSkillOptionItem(BaseModel):
    """A job's currently-required skill, with its id — unlike JobListItem.required_skills
    (a flat list of names via STRING_AGG), this is what the recruiter-edit skill picker
    needs to pre-fill selections and submit skill_ids back to PUT /jobs/{job_id}."""
    id: int
    name: str


class RerunAtsResponse(BaseModel):
    queued: int
    skipped_pending: int
    excluded: int
    in_progress_count: int
    not_stale_count: int
    message: str


class RerunAtsStatusResponse(BaseModel):
    """Polled by the recruiter dashboard while a rerun batch is in flight — `pending`
    is derived from how many of the applications queued by the most recent rerun still
    have an ats_evaluated_at older than when the batch was dispatched."""
    total_queued: int
    completed: int
    in_progress: bool


class JobInterviewRoundItem(BaseModel):
    round_order: int
    round_type_name: str
    failing_criteria: int | None
    description: str | None
    time_limit_minutes: int | None


class ATSCriterionSummary(BaseModel):
    section: str
    weight: float


class ATSCriteriaSummary(BaseModel):
    """Read-only, normalized view of a job's ats_criteria for display (e.g. the
    recruiter dashboard's job detail dialog). has_config=False means the job predates
    recruiter-defined ATS weighting (or the stored JSON was malformed) and the
    hardcoded default category weights are in effect instead of `criteria`.
    """

    has_config: bool
    criteria: list[ATSCriterionSummary]
    qualify_threshold: float
    overqualify_threshold: float | None
    auto_reject_overqualified: bool


class JobListItem(BaseModel):
    id: str
    description: str
    company: str
    job_role_id: int
    job_role_title: str
    experience_level_id: int
    experience_level_name: str
    location: str
    job_type: str
    salary_range: str | None
    posted_at: str
    expires_at: str | None
    required_skills: list[str]
    status: str = 'active'
    # Only populated for recruiter-facing listings (list_recruiter_jobs) — omitted (None)
    # from the public/candidate job list so scoring thresholds aren't exposed to candidates.
    ats_criteria: ATSCriteriaSummary | None = None


# ── Job analytics schemas ─────────────────────────────────────────────────────

class JobStatsRound(BaseModel):
    round_order: int
    round_type_name: str
    failing_criteria: int | None
    applicants_count: int


class JobStatsResponse(BaseModel):
    job_title: str
    description: str
    status: str
    required_skills: list[str]
    rounds: list[JobStatsRound]
    total_applicants: int
    passed_all_rounds: int
    hired_count: int


class RoundCandidateItem(BaseModel):
    candidate_id: str
    application_id: str
    interview_id: str
    name: str
    status: str


class CandidateSkillItem(BaseModel):
    name: str
    proficiency_level: str | None


class CandidateInfo(BaseModel):
    first_name: str
    last_name: str
    email: str
    bio: str | None
    current_location: str | None
    experience_level: str | None
    job_role: str | None
    skills: list[CandidateSkillItem]
    phone: str | None
    linkedin_url: str | None
    experience_years_min: int | None
    experience_years_max: int | None


class InterviewProgressItem(BaseModel):
    round_order: int
    round_type_name: str
    status: str | None
    result: float | None
    completed_at: str | None
    interview_id: str | None


class EvaluationQuestionItem(BaseModel):
    question_text: str
    candidate_answer: str | None
    score: int | None
    notes: str | None


class CandidatePanelResponse(BaseModel):
    candidate: CandidateInfo
    progress: list[InterviewProgressItem]
    evaluation: list[EvaluationQuestionItem] | None
